git.go

```diff
@@ -1,38 +1,268 @@
 package main
 
 import (
+	"fmt"
+	"os"
+	"time"
+
 	git "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/pkg/errors"
 	"go.opentelemetry.io/otel/attribute"
 )
 
 type GitScm struct {
+	baseRef        string
+	headSha        string
+	changeRequest  bool // if the tool is evaluating a change request or a branch
+	provider       string
+	repository     *git.Repository
 	repositoryPath string
 }
 
-// contributeOtelAttributes this method never fails, returning the current state of the contributed attributes
-// at the moment of the failure
-func (scm *GitScm) contributeOtelAttributes() []attribute.KeyValue {
-	repository, err := git.PlainOpen(scm.repositoryPath)
+func NewGitScm(repositoryPath string) *GitScm {
+	scm := &GitScm{
+		repositoryPath: repositoryPath,
+	}
+
+	repository, err := scm.openLocalRepository()
+	if err != nil {
+		return nil
+	}
+
+	scm.repository = repository
+
+	headSha, baseRef, gitProvider, changeRequest := checkGitProvider()
+
+	scm.headSha = headSha
+	scm.baseRef = baseRef
+	scm.changeRequest = changeRequest
+	scm.provider = gitProvider
+
+	return scm
+}
+
+// calculateCommits this method calculates the commits between current branch (HEAD) and a target branch.
+// - The target branch has to be set as the TARGET_BRANCH environment variable
+// - HEAD branch must be a valid branch in the git repository
+func (scm *GitScm) calculateCommits() (*object.Commit, *object.Commit, error) {
+	targetBranch, err := scm.repository.Branch(scm.baseRef)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "not able to retrieve the %s TARGET_BRANCH: %v", scm.baseRef, err)
+	}
+
+	targetRef, err := scm.repository.ResolveRevision(plumbing.Revision(targetBranch.Merge))
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "not able to retrieve ref from TARGET_BRANCH: %v", err)
+	}
+
+	targetCommit, err := scm.repository.CommitObject(*targetRef)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "not able to retrieve commit from TARGET_BRANCH: %v", err)
+	}
+
+	var headRefSha plumbing.Hash
+	if scm.headSha == "" {
+		headRef, err := scm.repository.Head()
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "not able to retrieve ref from HEAD: %v", err)
+		}
+
+		headRefSha = headRef.Hash()
+	} else {
+		headRefSha = plumbing.NewHash(scm.headSha)
+	}
+
+	headCommit, err := scm.repository.CommitObject(headRefSha)
 	if err != nil {
-		return []attribute.KeyValue{}
+		return nil, nil, errors.Wrapf(err, "not able to retrieve commit from HEAD: %v", err)
+	}
+
+	return headCommit, targetCommit, nil
+}
+
+// checkGitProvider identies the head sha and target branch from the environment variables that are
+// populated from a Git provider, such as Github or Gitlab. If no proprietary env vars are set, then it will
+// look up this tool-specific variable for the target branch.
+func checkGitProvider() (string, string, string, bool) {
```

Comment:
@v1v I'm addressing your concerns about running on different CI systems here. At this moment, only Gitlab and Github are supported. We could support Bitbucket, Gitea, Assembla, Azure DevOps, AWS CodeCommit, ....